linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IntegerDivisions.scala

```diff
@@ -0,0 +1,699 @@
+      } else {
+        // Keep looking
+        loop(p + 1, twoPowP << 1)
+      }
+    }
+
+    loop(p = W, twoPowP = twoPowW)
+  }
+
+  /** Like Integral[T], but with some unsigned operations that we need. */
+  sealed trait UnsignedIntegral[T] extends Integral[T] {
+    /** Number of bits used to represent a value of type `T`. */
+    val bitSize: Int
+
+    def isUnsignedPowerOf2OrZero(x: T): Boolean
+
+    /** Computes log2(x), assuming x is a power of 2. */
+    def log2Exact(x: T): Int
+
+    def toUnsignedBigInt(x: T): BigInt
+
+    def fromBigInt(x: BigInt): T
+
+    // IR-related operations
+
+    val tpe: Type
+
+    def literal(x: T)(implicit pos: Position): Literal
+
+    final def literalZero(implicit pos: Position): Literal = literal(zero)
+
+    // scalastyle:off disallow.space.before.token
+    val Op_+ : BinaryOp.Code
+    val Op_- : BinaryOp.Code
+    val Op_* : BinaryOp.Code
+    val Op_& : BinaryOp.Code
+    val Op_>>> : BinaryOp.Code
+    val Op_>> : BinaryOp.Code
+    // scalastyle:on disallow.space.before.token
+
+    def genMulSignedHi(x: T, y: VarRef, useRuntimeLong: Boolean)(implicit pos: Position): Tree
+    def genMulUnsignedHi(x: T, y: VarRef)(implicit pos: Position): Tree
+  }
+
+  implicit object IntIsUnsignedIntegral
+      extends UnsignedIntegral[Int]
+      with Numeric.IntIsIntegral with Ordering.IntOrdering {
+
+    val bitSize = 32
+
+    def isUnsignedPowerOf2OrZero(x: Int): Boolean =
+      (x & (x - 1)) == 0
+
+    def log2Exact(x: Int): Int =
+      31 - Integer.numberOfLeadingZeros(x)
+
+    def toUnsignedBigInt(x: Int): BigInt =
+      BigInt(Integer.toUnsignedLong(x))
+
+    def fromBigInt(x: BigInt): Int =
+      x.toInt
+
+    // IR-related operations
+
+    val tpe: Type = IntType
+
+    def literal(x: Int)(implicit pos: Position): Literal = IntLiteral(x)
+
+    // scalastyle:off disallow.space.before.token
+    val Op_+ : BinaryOp.Code = BinaryOp.Int_+
+    val Op_- : BinaryOp.Code = BinaryOp.Int_-
+    val Op_* : BinaryOp.Code = BinaryOp.Int_*
+    val Op_& : BinaryOp.Code = BinaryOp.Int_&
+    val Op_>>> : BinaryOp.Code = BinaryOp.Int_>>>
+    val Op_>> : BinaryOp.Code = BinaryOp.Int_>>
+    // scalastyle:on disallow.space.before.token
+
+    def genMulSignedHi(x: Int, y: VarRef, useRuntimeLong: Boolean)(
+        implicit pos: Position): Tree = {
+      /* (Math.multiplyFull(x, y) >>> 32).toInt
+       *
+       * Unfortunately, we cannot directly call that method, because it won't
+       * be available when we link a javalib < 1.20.
+       *
+       * Its user-land IR is easy enough to hard-code. However, we lose the
+       * handling as a RuntimeLong intrinsic done by the optimizer.
+       * That is why we special-case useRuntimeLong at this level, to directly
+       * emit a call to the intrinsic implementation.
+       *
+       * On the flip side, that allows the ElementaryInterpreter to easily
+       * handle this code (for useRuntimeLong = false), so it's not lost.
```

Comment:
I feel this sentence is quite hard to understand. Consider restructuring, maybe like so:

> On the flip side, when useRuntimeLong = false, it is easy for the ElementaryInterpreter to
> handle this code (only basic operations and literals required).